src/rigid.rs

```diff
@@ -0,0 +1,267 @@
+        Self {
+            rotation,
+            translation,
+        }
+    }
+
+    /// Construct an identity transform
+    #[inline]
+    pub fn identity() -> Self {
+        Self {
+            rotation: TypedRotation3D::identity(),
+            translation: TypedVector3D::zero(),
+        }
+    }
+
+    /// Construct a new rigid transformation, where the `translation` applies first
+    #[inline]
+    pub fn new_from_reversed(
+        translation: TypedVector3D<T, Src>,
+        rotation: TypedRotation3D<T, Src, Dst>,
+    ) -> Self {
+        // T * R
+        //   = (R * R^-1) * T * R
+        //   = R * (R^-1 * T * R)
+        //   = R * T'
+        //
+        // T' = (R^-1 * T * R) is also a translation matrix
+        // It is equivalent to the translation matrix obtained by rotating the
+        // translation by R
+
+        let translation = rotation.rotate_vector3d(&translation);
+        Self {
+            rotation,
+            translation,
+        }
+    }
+
+    #[inline]
+    pub fn from_rotation(rotation: TypedRotation3D<T, Src, Dst>) -> Self {
+        Self {
+            rotation,
+            translation: TypedVector3D::zero(),
+        }
+    }
+
+    #[inline]
+    pub fn from_translation(translation: TypedVector3D<T, Dst>) -> Self {
+        Self {
+            translation,
+            rotation: TypedRotation3D::identity(),
+        }
+    }
+
+    /// Decompose this into a translation and an rotation to be applied in the opposite order
+    ///
+    /// i.e., the translation is applied _first_
+    #[inline]
+    pub fn decompose_reversed(&self) -> (TypedVector3D<T, Src>, TypedRotation3D<T, Src, Dst>) {
+        // self = R * T
+        //      = R * T * (R^-1 * R)
+        //      = (R * T * R^-1) * R)
+        //      = T' * R
+        //
+        // T' = (R^ * T * R^-1) is T rotated by R^-1
+
+        let translation = self.rotation.inverse().rotate_vector3d(&self.translation);
+        (translation, self.rotation)
+    }
+
+    /// Returns the multiplication of the two transforms such that
+    /// other's transformation applies after self's transformation.
+    ///
+    /// i.e., this produces `self * other` in row-vector notation
+    #[inline]
+    pub fn post_mul<Dst2>(
+        &self,
+        other: &TypedRigidTransform3D<T, Dst, Dst2>,
+    ) -> TypedRigidTransform3D<T, Src, Dst2> {
+        // self = R1 * T1
+        // other = R2 * T2
+        // result = R1 * T1 * R2 * T2
+        //        = R1 * T1 * (R1^-1 * R1) * R2 * T2
+        //        = (R1 * T1 * R1^-1) * R1 * R2 * T2
+        //        = T' * R' * T2
+        //        = (R' * R'^-1) * T' * R' * T2
+        //        = R' * (R'^-1 * T' * R') * T2
+        //        = R' * T'' * T2
+        //        = R' * T''
+        //
+        // (R1 * T1 * R1^-1) = T' = T1 rotated by R1^-1
+        // R1 * R2  = R'
+        // (R'^-1 * T' * R') = T'' = T' rotated by R'
+        // T'' * T2 = T''' = T2 + T'
```

Comment:
Also I don't understand this line either, sorry…

Reply:
clarified

Reply:
That doesn't seem to match the code below, which has `T''' = T'' + T2`, not `T''' = T' + T2` as you wrote there.

Reply:
Right, this diff is outdated, look at it from the top